atomic_reactor/operator_util.py

```diff
@@ -0,0 +1,233 @@
+"""
+Copyright (c) 2020 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+
+from __future__ import absolute_import, unicode_literals
+
+import os
+import logging
+
+from ruamel.yaml import YAML
+from ruamel.yaml.comments import CommentedMap, CommentedSeq
+
+from atomic_reactor.util import ImageName, chain_get
+
+
+yaml = YAML()
+log = logging.getLogger(__name__)
+
+
+OPERATOR_CSV_KIND = "ClusterServiceVersion"
+
+
+class OperatorCSV(object):
+    """
+    A single ClusterServiceVersion file in an operator manifest.
+
+    Can find and replace pullspecs for container images in predefined locations.
+    """
+
+    def __init__(self, path, data):
+        """
+        Initialize an OperatorCSV
+
+        :param path: Path where data was found or where it should be written
+        :param data: ClusterServiceVersion yaml data
+        """
+        if data.get("kind") != OPERATOR_CSV_KIND:
+            raise ValueError("Not a ClusterServiceVersion")
+        self.path = path
+        self.data = data
+
+    @classmethod
+    def from_file(cls, path):
+        """
+        Make an OperatorCSV from a file
+
+        :param path: Path to file
+        :return: OperatorCSV
+        """
+        with open(path) as f:
+            data = yaml.load(f)
+            return cls(path, data)
+
+    def write(self):
+        """
+        Write data to file (preserves comments)
+        """
+        with open(self.path, "w") as f:
+            yaml.dump(self.data, f)
+
+    def get_pullspecs(self):
+        """
+        Find pullspecs for containers, related images and RELATED_IMAGE_*
+        environment variables in containers.
+
+        :return: set of ImageName pullspecs
+        """
+        pullspecs = self._get_related_image_pullspecs()
+        pullspecs.update(self._get_env_pullspecs())
+        pullspecs.update(self._get_container_image_pullspecs())
+        return pullspecs
+
+    def replace_pullspecs(self, replacement_pullspecs):
+        """
+        Replace pullspecs for containers, related images and RELATED_IMAGE_*
+        environment variables in containers.
+
+        :param replacement_pullspecs: mapping of pullspec -> replacement
+        """
+        self._replace_related_image_pullspecs(replacement_pullspecs)
+        self._replace_env_pullspecs(replacement_pullspecs)
+        self._replace_container_image_pullspecs(replacement_pullspecs)
+
+    def replace_pullspecs_everywhere(self, replacement_pullspecs):
+        """
+        Replace all pullspecs found anywhere in data
+
+        :param replacement_pullspecs: mapping of pullspec -> replacment
+        """
+        log_template = "{path} - Replaced pullspec: {old} -> {new}"
```

Comment:
Why? Can we have just constant `LOG_TEMPLATE = "%s - Replaced pullspec: %s --> %s"` without interpolation adn conditions?

Reply:
I thought it would be nice to log where exactly a pullspec was found/replaced, but I did manage to get rid of string interpolation FWIW